### PR TITLE
[DO NOT MERGE] Add support for fully matrix-free Mass and Diffusion operators using libCEED

### PR DESCRIPTION
**Pull request by YohannDudouit**
This PR adds fully matrix-free operators using libCEED for the Mass and Diffusion Integrators. It also adds matrix-free diagonal assembly using @jeremylt 's work.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1670](https://github.com/mfem/mfem/pull/1670) | @YohannDudouit | @v-dobrev | @pazner + @delyank | 08/20/20 | 09/10/20 | 09/10/20 | |

### CODE REVIEW

**Comment by YohannDudouit** [comment]
@pazner @delyank You can test this feature by modifying `ex1`, you need to use `SetAssemblyLevel(AssemblyLevel::NONE);` on the `BilinearForm`.

File: fem/libceed/ceed.cpp
Comment (pazner) [inline fem/libceed/ceed.cpp:604]:
```suggestion
   CeedQFunctionSetContext(ceedData.apply_qfunc, ceedData.build_ctx);
```

I think it will segfault because of this 😱

File: fem/libceed/mass.h
Comment (YohannDudouit) [reply fem/libceed/mass.h:152]: I was thinking to rename this to `f_apply_mass_mf_quad` once #1716 is merged. I will also add the vectorial versions when #1713 is merged.

File: fem/libceed/ceed.hpp
Comment (YohannDudouit) [reply fem/libceed/ceed.hpp:184]: #1716

**Comment by YohannDudouit** [comment]
@pazner I added a constructor to `CeedData` to initialize the fields to `nullptr`.

File: fem/bilinearform_ext.hpp
Comment (delyank) [inline fem/bilinearform_ext.hpp:133]: You may want to remove the "NOT YET IMPLEMENTED" statement.

**Comment by delyank** [comment]
> @pazner @delyank You can test this feature by modifying `ex1`, you need to use `SetAssemblyLevel(AssemblyLevel::NONE);` on the `BilinearForm`.

Do you plan to add the `SetAssemblyLevel(AssemblyLevel::NONE);` officially in the examples (particularly, `ex1`), probably using another flag and parameter (like `-mf` and `bool mf;`)? Is it supposed to be done in a separate PR?

At a glance, I like the PR and the code looks clean and it seems that it adds only (or almost only) new functionality without modifying existent functionality. Apparently, this is related and interacts with other PRs.

I would like to look at it and play with it a bit more.

**Comment by YohannDudouit** [comment]
@delyank The is a recurring request, but `SetAssemblyLevel(AssemblyLevel::NONE)` will not be in `ex1` to keep it simple. (We were wondering also for `AssemblyLevel::ELEMENT` and `AssemblyLevel::FULL`)
@delyank @pazner If you don't have concerns about the style/design, I can merge it to #1719 and test it there.

File: fem/libceed/ceed.hpp
Comment (delyank) [inline fem/libceed/ceed.hpp:221]: You removed "PA" from the names of the functions, which I guess is because you have a more general behavior now. If this is the case, you should remove the "PA" from the comments.

**Comment by delyank** [comment]
> @delyank The is a recurring request, but `SetAssemblyLevel(AssemblyLevel::NONE)` will not be in `ex1` to keep it simple. (We were wondering also for `AssemblyLevel::ELEMENT` and `AssemblyLevel::FULL`)

OK. So the discussion on the examples is separate from this PR.

> @delyank @pazner If you don't have concerns about the style/design, I can merge it to #1719 and test it there.

I have no concerns about the design or style. You will need to resolve the merging anyway once things get merged in next one by one.

**Comment by YohannDudouit** [comment]
@delyank The idea is that I merge everything in #1719 so that we have only one PR to merge into next at the end.

**Comment by tzanio** [comment]
> @delyank The idea is that I merge everything in #1719 so that we have only one PR to merge into next at the end.

I am confused -- should this review continue or not?

**Comment by YohannDudouit** [comment]
@tzanio Sure, I'm just making it easier to merge in `next` by solving all the merge conflicts ahead in #1719 .

**Comment by delyank** [comment]
> @tzanio Sure, I'm just making it easier to merge in `next` by solving all the merge conflicts ahead in #1719 .

I think I understand. You mean that when you merge this into #1719, then #1719 can only be approved after this one is approved, since #1719 is like a PR against this current PR. That is, it is like extending on top of this PR.

**Comment by YohannDudouit** [comment]
@pazner @delyank Once you've reviewed and approved this PR, could you please check that the merge in #1719 looks good to you, and if it does write "The merge of #1670 in #1719 looks good to me."
I will close this PR afterwards in favor of #1719 to avoid complex merge conflicts for the editors.

File: fem/libceed/ceed.hpp
Comment (YohannDudouit) [reply fem/libceed/ceed.hpp:221]: done.

**Review by delyank — APPROVED** [review]
I'm generally happy with this code. Let it be merged into #1719.

**Comment by YohannDudouit** [comment]
@pazner @delyank Once you've reviewed and approved this PR, could you please check that the merge in #1719 looks good to you, and if it does write "The merge of #1670 in #1719 looks good to me."
I will close this PR afterwards in favor of #1719 to avoid complex merge conflicts for the editors.

**Comment by pazner** [comment]
I think there are still some issues with the destructor `~CeedData()` (for example, with fully matrix-free, `build_oper` won't be created but will be destroyed).

**Comment by jeremylt** [comment]
So long as the pointer to the build operator is set to `NULL` in the matrix free case, it should be ok. It is safe to pass `NULL` to a libCEED destructor.

**Comment by pazner** [comment]
> So long as the pointer to the build operator is set to `NULL` in the matrix free case, it should be ok. It is safe to pass `NULL` to a libCEED destructor.

Currently these pointers are uninitialized, so they may not be NULL. Looks like these pointers are NULL-initialized in #1713, so this will be fixed after the PRs are merged into #1719.

I want to make sure this fix doesn't fall through the cracks in the merge process.

**Comment by YohannDudouit** [comment]
@pazner I think this is fixed in #1719

**Review by pazner — APPROVED** [review]

**Comment by v-dobrev** [comment]
Okay: this PR is now approved and merged into https://github.com/mfem/mfem/tree/yohann/test-ceed, so I'll mark it as approved and merged in the table in the first comment. Then, for clarity, I'll switch the target branch to be https://github.com/mfem/mfem/tree/yohann/test-ceed and that should automatically change the state of the PR to "merged", I think.

**Comment by v-dobrev** [comment]
GitHub is not letting me switch the target branch to 'yohann/test-ceed' because:
```
There are no new commits between base branch 'yohann/test-ceed' and head branch 'yohann/ceed-mf'
```
Because of that, I'll simply close this PR.

**Pull request closed without merging**
